src/psbt.c

```diff
@@ -0,0 +1,2188 @@
+        }
+        input->unknowns = result_unknowns;
+    }
+    input->sighash_type = sighash_type;
+
+    return ret;
+
+fail:
+    if (result_non_witness_utxo) {
+        wally_tx_free(result_non_witness_utxo);
+    }
+    if (result_witness_utxo) {
+        wally_tx_output_free(result_witness_utxo);
+    }
+    if (result_redeem_script) {
+        wally_free(result_redeem_script);
+    }
+    if (result_witness_script) {
+        wally_free(result_witness_script);
+    }
+    if (result_final_script_sig) {
+        wally_free(result_final_script_sig);
+    }
+    if (result_final_witness) {
+        wally_tx_witness_stack_free(result_final_witness);
+    }
+    if (result_keypaths) {
+        keypath_map_free(result_keypaths);
+    }
+    if (result_partial_sigs) {
+        partial_sigs_map_free(result_partial_sigs);
+    }
+    if (result_unknowns) {
+        unknowns_map_free(result_unknowns);
+    }
+    return ret;
+}
+
+int wally_psbt_input_free(struct wally_psbt_input *input)
+{
+    if (input) {
+        clear_and_free(input->redeem_script, input->redeem_script_len);
+        clear_and_free(input->witness_script, input->witness_script_len);
+        clear_and_free(input->final_script_sig, input->final_script_sig_len);
+        wally_tx_witness_stack_free(input->final_witness);
+        keypath_map_free(input->keypaths);
+        partial_sigs_map_free(input->partial_sigs);
+        unknowns_map_free(input->unknowns);
+    }
+    return WALLY_OK;
+}
+
+int wally_psbt_output_init_alloc(
+    unsigned char *redeem_script,
+    size_t redeem_script_len,
+    unsigned char *witness_script,
+    size_t witness_script_len,
+    struct keypath_map *keypaths,
+    struct unknowns_map *unknowns,
+    struct wally_psbt_output **output)
+{
+    struct wally_psbt_output *result;
+    int ret = WALLY_OK;
+
+    TX_CHECK_OUTPUT;
+    TX_OUTPUT_ALLOC(struct wally_psbt_output);
+
+    ret = wally_psbt_output_update(result,
+                                   redeem_script,
+                                   redeem_script_len,
+                                   witness_script,
+                                   witness_script_len,
+                                   keypaths,
+                                   unknowns);
+
+    if (ret != WALLY_OK) {
+        clear_and_free(result, sizeof(*result));
+        *output = NULL;
+    }
+    return ret;
+}
+
+int wally_psbt_output_update(
+    struct wally_psbt_output *output,
+    unsigned char *redeem_script,
+    size_t redeem_script_len,
+    unsigned char *witness_script,
+    size_t witness_script_len,
+    struct keypath_map *keypaths,
+    struct unknowns_map *unknowns)
+{
+    unsigned char *result_redeem_script = NULL;
+    unsigned char *result_witness_script = NULL;
+    struct keypath_map *result_keypaths = NULL;
+    struct unknowns_map *result_unknowns = NULL;
+    int ret = WALLY_OK;
+
+    if (redeem_script && !clone_bytes(&result_redeem_script, redeem_script, redeem_script_len)) {
```

Comment:
General comment on this function - it may be the case that output already has buffers allocated that are big enough to hold the new redeem_script, witness_script, keypaths and unknowns, in which case it would be better to reuse them rather than freeing the memory and allocating new.

Reply:
With how the rest of the functions are written, it's necessary for the buffers to be the exact length of the scripts otherwise too much stuff may be written out. So even if the buffer is big enough, it may not be a good idea to use it.